### PR TITLE
BHBC-1188: Timezone bug fix

### PR DESCRIPTION
**Pull request by NickPhura**
# Overview

## Links to jira tickets

https://quartech.atlassian.net/browse/BHBC-1188

## This PR contains the following changes

- Add function in `db.ts` to stop node-postgres from adding default time/timezone information to psql `DATE` values (which by design don't contain time/timezone information).

Bonus:
- Update docker-compose/env/pipeline files for Db and API to allow a timezone to be specified (rather than just letting it default to whatever default value, UTC probably).

## This PR contains the following types of changes

- [ ] New feature (change which adds functionality)
- [x] Enhancement (improvements to existing functionality)
- [x] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [x] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [x] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [x] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [x] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Tested locally.

Also used the PR-deploy to test that the timezone for API and DB do change if I change the config.json timezone values.

### CODE REVIEW

**Review by rstens — COMMENTED** [review]
So the standard TZ on the pod is already: ENV TZ America/Vancouver

I am a bit confused why this has to be reset.

**Comment by NickPhura** [comment]
> 
> 
> So the standard TZ on the pod is already: ENV TZ America/Vancouver
> 
> I am a bit confused why this has to be reset.

Ya I kind of combined the actual change needed, with some bonus stuff.

The actual change that fixed some of the funny app dates is just a few lines in `db.ts`.

Everything else is just bonus to give us the capability of specifying the timezones used by the API and DB, in case we ever want to change it. A bunch of this came about naturally as part of testing the timezones (changing the API/DB to various timezone combinations to see how it affected the dates, etc), and so I just carried it forward into the pipeline since it wasn't a big additional change.

But ya, you can control the timezones individually for both api and db now in the local docker setup, and optionally change that root config.json if we ever wanted to have our openshift environments deployed in a different timezone (UTC, or w/e).

File: api/src/database/db.ts [hunk 1]
```diff
@@ -29,10 +29,19 @@ const poolConfig: PoolConfig = {
 
 defaultLog.debug({ label: 'create db pool', message: 'pool config', poolConfig });
 
-let pool: Pool;
+// Custom type handler for psq `DATE` type to prevent local time/zone information from being added.
+// Why? By default, node-postgres assumes local time/zone for any psql `DATE` or `TIME` types that don't have timezone information.
+// This Can lead to unexpected behaviour when the original psql `DATE` value was intentionally omitting time/zone information.
+// PSQL date types: https://www.postgresql.org/docs/12/datatype-datetime.html
+// node-postgres type handling (see bottom of page): https://node-postgres.com/features/types
+pg.types.setTypeParser(1082, (stringValue: string) => {
```
Comment (anissa-agahchen) [inline api/src/database/db.ts]:
```suggestion
pg.types.setTypeParser(pg.types.builtins.DATE, (stringValue: string) => {
```

Comment (anissa-agahchen) [reply api/src/database/db.ts]: Do we need to set a similar typeparser for timestamp and timestampz?

Comment (NickPhura) [reply api/src/database/db.ts]: oh nice one! I didn't realize there was a built in variable for that.

Right now its only `DATE` that we have issues with.  `timestamp(z)` already includes time/timezone information, so there is no issues as node-postgres doesn't alter that in any way.  Its only an issue for the types that don't have time/timezone information, as node-postgres decides to add it anyways, using some default values.  

I didn't add a function for `TIME`, which is the other type that node-postgres messes with, as we dont have any UI that creates time values yet.  We really only have a handful of date fields in the whole app and all of them are just start/end dates, with no time component.

**Review by anissa-agahchen — APPROVED** [review]

**Pull request merged**